pytensor/tensor/rewriting/linalg.py

```diff
@@ -569,3 +569,45 @@ def svd_uv_merge(fgraph, node):
                     or len(fgraph.clients[cl.outputs[2]]) > 0
                 ):
                     return [cl.outputs[1]]
+
+
+@register_canonicalize
+@register_stabilize
+@node_rewriter([Blockwise])
+def rewrite_inv_inv(fgraph, node):
+    """
+    This rewrite takes advantage of the fact that if there are two consecutive inverse operations (inv(inv(input))), we get back our original input without having to compute inverse once.
+
+    Here, we check for direct inverse operations (inv/pinv)  and allows for any combination of these "inverse" nodes to be simply rewritten.
+
+    Parameters
+    ----------
+    fgraph: FunctionGraph
+        Function graph being optimized
+    node: Apply
+        Node of the function graph to be optimized
+
+    Returns
+    -------
+    list of Variable, optional
+        List of optimized variables, or None if no optimization was performed
+    """
+    valid_inverses = (MatrixInverse, MatrixPinv)
+    # Check if its a valid inverse operation (either inv/pinv)
+    # In case the outer operation is an inverse, it directly goes to the next step of finding inner operation
+    # If the outer operation is not a valid inverse, we do not apply this rewrite
+    if not isinstance(node.op.core_op, valid_inverses):
+        return None
```

Comment:
If we predefine the two possible pinv as blockwise (like we do for matrix_inverse) the rewrite can track more specifically and avoids being called for any Blockwise it sees:


```suggestion
@node_rewriter([matrix_inverse, matrix_pinv_hermitian, matrix_pinv_non_hermitian])
def rewrite_inv_inv(fgraph, node):
    """
    This rewrite takes advantage of the fact that if there are two consecutive inverse operations (inv(inv(input))), we get back our original input without having to compute inverse once.

    Here, we check for direct inverse operations (inv/pinv)  and allows for any combination of these "inverse" nodes to be simply rewritten.

    Parameters
    ----------
    fgraph: FunctionGraph
        Function graph being optimized
    node: Apply
        Node of the function graph to be optimized

    Returns
    -------
    list of Variable, optional
        List of optimized variables, or None if no optimization was performed
    """
```

Need to predefine those matrix_pinv*.. The helper "pinv" should return the predefined Ops instead of creating new ones to avoid Op duplication

Reply:
i dont quite understand the last line u wrote about 
> Need to predefine those matrix_pinv*.. The helper "pinv" should return the predefined Ops instead of creating new ones to avoid Op duplication
also, as a general rule for rewrites, is it better that they should be tracking more specific Ops instead of just Blockwise

Reply:
> also, as a general rule for rewrites, is it better that they should be tracking more specific Ops instead of just Blockwise

Is that a question? The answer is yes. It avoids useless calls to the rewrite function when the Op is not in the graph

Reply:
but for the check of the ops inside, i will have to use the method that i am already doing

Reply:
Regarding pinv, the helper function is here: https://github.com/pymc-devs/pytensor/blob/ad27dc752798dfb75ed8d1c0a9b840859a8a898d/pytensor/tensor/nlinalg.py#L68

The idea is instead:

```python

pinv_hermitian = Blockwise(Pinv(hermitian=True))

pinv_non_hermitian = Blockwise(Pinv(hermitian=False))

def pinv(x, hermitian=False):
  ...
  return pinv_hermitian if hermitian else pinv_non_hermitian

Reply:
But actually let's not to that now, have to think. This makes initialization a bit slower because we have to create more instances...

Reply:
oh alright. but this is essentially the same thing as first checking the op as blockwise and then the core op as one of the valid inverses. whats the difference in both these ideas

Reply:
The difference is the rewrite is not even considered if the node is not a Blockwise(Pinv). This way the rewrite is attempted on every single Blockwise in the graph. 

It's just an optimization, not a question of correctness